mysql/infoschema.go

```diff
@@ -0,0 +1,367 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"database/sql"
+	"fmt"
+	"strings"
+
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+	_ "github.com/go-sql-driver/mysql" // The driver should be used via the database/sql package.
+	_ "github.com/lib/pq"
+)
+
```

Comment:
There is significant overlap between mysql/infoschema.go and postgres/inforschema.go.

Can you summarize the differences?

Probably worth a TODO to capture this.

Reply:
Main differences between postgres/infoschema.go and mysql/infoschema.go:

- SQL statements.
   - Difference in sql statements `ProcessSQLData` and `getColumns`.
   - In `SetRowStats` -  dropped the `" "` from table_schema and table_name as per mysql syntax.

- Different way to read data i.e. instead of interface{} we are using sql.RawBytes{}.

- Different ways of handling data conversion. Instead of `ConvertSQLRow`,`cvtSQLScalar` and `cvtSQLArray` function used for data conversion in postgres, mysql is converting row values to string and then using `ProcessDataRow` function of `mysql/data.go` file for data conversion. Hence, removed the redundancy of data conversion functions by directly using functions in `mysql/data.go`.

- Added a function `buildColNameList` to handle spatial datatypes. 

- Different list of system tables to be ignored in `getTables` function. 
   - Mysql - `{"information_schema", "mysql", "performance_schema", "sys"}` 
   - Postgres - `{"information_schema", "postgres", "pg_catalog", "pg_temp_1", "pg_toast", "pg_toast_temp_1"}`

- Do not need to drop `public` prefix in table name because mysql do not support `public`.

- `valsToStrings` is different because postgres uses `[]interface{}` argument, while mysql uses `[]sql.RawBytes` argument.

Reply:
Many thanks for the detailed response.

What's your take on RawBytes vs interface{}?

My understanding was that interface{} would return values from the underlying driver without conversion.
But RawBytes effectively forces conversion to a string format. Which means that if the value is an int64, then we do int64 -> string -> int64.

But I 100% agree with you about the redundant data conversion code.

Should we just do both postgres and mysql via string? Any thoughts on the additional processing/conversion cost?

Reply:
Main benefit to use RawBytes is to avoid redundancy in code and reuse data conversion functions. Other benefit is that we can avoid complex switch conditions in `convScalar` and `convArray`.
It would definitely be easier and in sync to use RawBytes in postgres as well. [This](github.com/lib/pq) library is currently used in postgres. Need to check if it supports RawBytes. Also, in the documentation of current library another library([LINK](https://github.com/jackc/pgx)) is recommended to be used. In addition to that,array structure is a bit complex in postgres compared to mysql need to check that also.

We can discuss more about additional processing/conversion cost on google chat group.